src/plugins/tool_drag.py

```diff
@@ -32,11 +32,80 @@ def __init__(self, api):
         self.action.setShortcut(QtGui.QKeySequence("Ctrl+4"))
         # Register the tool
         self.api.register_tool(self)
+        self._stamp = []
+        self._lastdraw = []
+        self._original = []
+        self.xdir = True
+        self.ydir = True
+        self.zdir = True
+        self.pastoffset = 0
+        self.fixeddirection = False
+        self.keeporiginal = False
+        self._mouse = ()
+        self.drawlock = threading.Lock()
+
+    def check_free_space(self, data, dx, dy, dz):
+        for x, y, z, col in self._stamp:
+            newx = (x + dx) % data.voxels.width
+            newy = (y + dy) % data.voxels.height
+            newz = (z + dz) % data.voxels.depth
+            if data.voxels.get(newx, newy, newz) != 0 and not (newx, newy, newz) in data.voxels._selection:
+                return False
+        return True
+
+    def drawstamp(self, data, dx, dy, dz, keeporiginal=False):
+        self.drawlock.acquire()
+        if self.check_free_space(data, dx, dy, dz):
+            for x, y, z in self._lastdraw:
+                if not keeporiginal or not (x, y, z) in self._original:
+                    data.voxels.set(x, y, z, 0, True, 1)
+            self._lastdraw = []
+            data.voxels.clear_selection()
+            for x, y, z, col in self._stamp:
+                newx = (x + dx) % data.voxels.width
+                newy = (y + dy) % data.voxels.height
+                newz = (z + dz) % data.voxels.depth
+                data.voxels.set(newx, newy, newz, col, True, 1)
```

Comment:
I changed drawstamp so that it doesn't set voxels twice. The bug still persists. Every once in a while voxels don't get deleted.